optimizely/client/client.go

```diff
@@ -364,43 +360,107 @@ func (o *OptimizelyClient) GetAllFeatureVariables(featureKey string, userContext
+	defer func() {
+		if r := recover(); r != nil {
+			switch t := r.(type) {
+			case error:
+				err = t
+			case string:
+				err = errors.New(t)
+			default:
+				err = errors.New("unexpected error")
+			}
+			errorMessage := fmt.Sprintf("optimizely SDK is panicking with the error:")
+			logger.Error(errorMessage, err)
+			logger.Debug(string(debug.Stack()))
+		}
+	}()
+
+	_, experimentDecision, err := o.getExperimentDecision(experimentKey, userContext)
+	if err != nil {
+		logger.Error("received an error while computing experiment decision", err)
+	}
+
+	if experimentDecision.Variation != nil {
+		result = experimentDecision.Variation.Key
+	}
+
+	return result, err
+}
+
+// Track take and event key with event tags and if the event is part of the config, send to events backend.
+func (o *OptimizelyClient) Track(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}) (err error) {
 
 	defer func() {
-		var e error
 		if r := recover(); r != nil {
 			switch t := r.(type) {
 			case error:
-				e = t
+				err = t
 			case string:
-				e = errors.New(t)
+				err = errors.New(t)
 			default:
-				e = errors.New("unexpected error")
+				err = errors.New("unexpected error")
 			}
 			errorMessage := fmt.Sprintf("optimizely SDK is panicking with the error:")
-			logger.Error(errorMessage, e)
+			logger.Error(errorMessage, err)
 			logger.Debug(string(debug.Stack()))
+		}
+	}()
+
+	projectConfig, err := o.GetProjectConfig()
+	if err != nil {
+		logger.Error("Optimizely SDK tracking error", err)
+		return err
+	}
+
+	configEvent, err := projectConfig.GetEventByKey(eventKey)
+
+	if err == nil {
+		userEvent := event.CreateConversionUserEvent(projectConfig, configEvent, userContext, eventTags)
+		o.eventProcessor.ProcessEvent(userEvent)
+	} else {
+		errorMessage := fmt.Sprintf(`optimizely SDK track: error getting event with key "%s"`, eventKey)
+		logger.Error(errorMessage, err)
+		return err
+	}
 
-			// If we have a feature, then we can recover w/o throwing
-			if decisionContext.Feature == nil {
-				err = e
+	return err
+}
+
+func (o *OptimizelyClient) getFeatureDecision(featureKey string, userContext entities.UserContext) (decisionContext decision.FeatureDecisionContext, featureDecision decision.FeatureDecision, err error) {
+
+	defer func() {
+		if r := recover(); r != nil {
+			switch t := r.(type) {
+			case error:
+				err = t
+			case string:
+				err = errors.New(t)
+			default:
+				err = errors.New("unexpected error")
 			}
+			errorMessage := fmt.Sprintf("optimizely SDK is panicking with the error:")
+			logger.Error(errorMessage, err)
+			logger.Debug(string(debug.Stack()))
 		}
 	}()
 
 	userID := userContext.ID
 	logger.Debug(fmt.Sprintf(`Evaluating feature "%s" for user "%s".`, featureKey, userID))
 
-	projectConfig, err := o.GetProjectConfig()
-	if err != nil {
-		logger.Error("Error calling getFeatureDecision", err)
-		return decisionContext, featureDecision, err
+	projectConfig, e := o.GetProjectConfig()
```

Comment:
Because we already named the return error parameter `err` and we don't want to re-assign it here. Here is an article that explains it: https://yourbasic.org/golang/gotcha-shadowing-variables/